src/lesson-1-9-demo/SignupForm.jsx

```diff
@@ -1,23 +1,165 @@
 import { useEffect, useState } from "react";
 import "./signup-form.css";
 
+const createUser = async (userData) => {
+  try {
+    const response = await fetch("http://localhost:3001/api/users", {
+      headers: {
+        Accept: "application/json",
+        "Content-Type": "application/json",
+      },
+      method: "POST",
+      body: JSON.stringify(userData),
+    });
+
+    if (response.ok) {
+      return response.json();
+    }
+
+    throw Error("Failed to create user");
+  } catch (error) {
+    console.error({ error });
+    throw error;
+  }
+};
+
+const initialFormState = {
+  username: "",
+  email: "",
+  password: "",
+  passwordConfirmation: "",
+};
+
 const SignupForm = () => {
+  const [formState, setFormState] = useState(initialFormState);
+  const [formErrors, setFormErrors] = useState({});
+
+  const validations = [
+    {
+      key: "username",
+      isValid: (formState) => formState.username.trim().length > 0,
+      errorMessage: "username cannot be blank",
+    },
+    {
+      key: "email",
+      isValid: (formState) =>
+        formState.email.length > 0 &&
+        /^[\w-\.]+@([\w-]+\.)+[\w-]{2,4}$/.test(formState.email),
+      errorMessage: "email is not valid",
+    },
+    {
+      key: "password",
+      isValid: (formState) => {
+        return (
+          formState.password.length > 12 &&
```

Comment:
we define a series of password checks here:
- check if the password is at least 13 characters
- check if the password has a lowercase letter
- check if the form has an uppercase latter
- check if the form has a number
- check if the form has a symbol